File: example/example_double_im6014.cpp
#include "unitree_im6014/im6014_motor.hpp"
#include <atomic>
#include <chrono>
#include <csignal>
#include <cmath>
#include <iostream>
#include <thread>

using namespace unitree::IM6014;

static std::atomic<bool> keep_running{true};

static void on_sigint(int) { keep_running.store(false); }

int main() {
  Motor motor;
  if (!motor.init("/dev/ttyUSB0", 6000000)) {
    std::cerr << "Failed to open serial port!" << std::endl;
    return -1;
  }

  std::signal(SIGINT, on_sigint);

  constexpr uint8_t id1 = 0, id2 = 1;
  constexpr float kp = 0.0f, kd = 2.5f;

  std::cout << "=== IM6014 Dual Motor Example (Output-end units) ==="
            << std::endl;

  State s1, s2;
  float t = 0.0f;
  while (keep_running.load()) {
    // Velocity control with sinusoidal trajectory (output-end units)
    motor.send_cmd(id1, 0.0f,
                   -0.5f * (1.0f - std::cos(3.0f * t)), 0.0f, kp, kd);
    motor.send_cmd(id2, 0.0f,
                    1.5f * (1.0f - std::cos(2.0f * t)), 0.0f, kp, kd);

    motor.poll_states(2);
    bool ok1 = motor.get_state(id1, s1);
    bool ok2 = motor.get_state(id2, s2);

    if (ok1 && ok2) {
      std::cout << "\033c" << std::flush;
      std::cout << "[M1] Pos: " << s1.pos << " rad, "
                << "Spd: " << s1.speed << " rad/s, "
                << "Tor: " << s1.torque << " Nm" << std::endl
                << "[M2] Pos: " << s2.pos << " rad, "
                << "Spd: " << s2.speed << " rad/s, "
                << "Tor: " << s2.torque << " Nm" << std::endl;
    }
    std::this_thread::sleep_for(std::chrono::milliseconds(2));
    t += 0.002f;
  }

  // Graceful stop
  std::cout << "\nCaught Ctrl+C, stopping motors..." << std::endl;
  motor.send_cmd(id1, 0, 0, 0, 0, 0, MotorStatus::DISABLE);
  motor.send_cmd(id2, 0, 0, 0, 0, 0, MotorStatus::DISABLE);
  std::this_thread::sleep_for(std::chrono::seconds(1));

  motor.close();
  std::cout << "Motors stopped." << std::endl;
  return 0;
}

File: example/example_double_im6014.py
#!/usr/bin/env python3
"""Dual IM6014 motor velocity-control example — matches example_double_im6014.cpp."""

import math
import os
import signal
import sys
import time

try:
    from unitree_im6014 import Motor, State
except ModuleNotFoundError:
    _project_root = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
    sys.path.insert(0, os.path.join(_project_root, "python"))
    from unitree_im6014 import Motor, State  # type: ignore[no-redef]

keep_running = True


def _on_sigint(_signum, _frame):
    global keep_running
    keep_running = False


def main():
    global keep_running
    motor = Motor()
    if not motor.init("/dev/ttyUSB0", 6000000):
        print("Failed to open serial port!", file=sys.stderr)
        return 1

    signal.signal(signal.SIGINT, _on_sigint)

    id1, id2 = 0, 1
    kp, kd = 0.0, 2.5

    print("=== IM6014 Dual Motor Example (Output-end units) ===")

    s1, s2 = State(), State()
    t = 0.0
    while keep_running:
        # Velocity control with sinusoidal trajectory (output-end units)
        motor.send_cmd(id1, 0.0, -0.5 * (1 - math.cos(3 * t)), 0, kp, kd)
        motor.send_cmd(id2, 0.0, 1.5 * (1 - math.cos(2 * t)), 0, kp, kd)

        motor.poll_states(timeout_ms=2)
        ok1 = motor.get_state(id1, s1)
        ok2 = motor.get_state(id2, s2)

        if ok1 and ok2:
            sys.stdout.write("\033c")
            sys.stdout.write(
                f"[M1] Pos: {s1.pos:.3f} rad, "
                f"Spd: {s1.speed:.3f} rad/s, "
                f"Tor: {s1.torque:.3f} Nm\n"
                f"[M2] Pos: {s2.pos:.3f} rad, "
                f"Spd: {s2.speed:.3f} rad/s, "
                f"Tor: {s2.torque:.3f} Nm\n"
            )
            sys.stdout.flush()

        time.sleep(0.002)
        t += 0.002

    # Graceful stop
    print("\nCaught Ctrl+C, stopping motors...")
    motor.send_cmd(id1, 0, 0, 0, 0, 0, status=0)
    motor.send_cmd(id2, 0, 0, 0, 0, 0, status=0)
    time.sleep(1.0)

    motor.close()
    print("Motors stopped.")
    return 0


if __name__ == "__main__":
    sys.exit(main())
